plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/AwsS3ServiceImplTests.java

```diff
@@ -0,0 +1,359 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cloud.aws;
+
+import com.amazonaws.AmazonClientException;
+import com.amazonaws.ClientConfiguration;
+import com.amazonaws.Protocol;
+import com.amazonaws.auth.AWSCredentials;
+import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.auth.EnvironmentVariableCredentialsProvider;
+import com.amazonaws.auth.InstanceProfileCredentialsProvider;
+import com.amazonaws.auth.SystemPropertiesCredentialsProvider;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.repositories.RepositorySettings;
+import org.elasticsearch.repositories.s3.S3Repository;
+import org.elasticsearch.test.ESTestCase;
+
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+
+public class AwsS3ServiceImplTests extends ESTestCase {
+
+    public void testAWSCredentialsWithSystemProviders() {
+        AWSCredentials expectedCredentials = null;
+
+        try {
+            expectedCredentials = new EnvironmentVariableCredentialsProvider().getCredentials();
+        } catch (AmazonClientException e1) {
+            // We don't have env variables
+            try {
+                expectedCredentials = new SystemPropertiesCredentialsProvider().getCredentials();
+            } catch (AmazonClientException e2) {
+                // We don't have system properties
+                try {
+                    expectedCredentials = new InstanceProfileCredentialsProvider().getCredentials();
+                } catch (AmazonClientException ignored) {
+                    // We don't have instance profile metadata
+                }
+            }
+        }
+
+        RepositorySettings repositorySettings = new RepositorySettings(Settings.EMPTY, Settings.EMPTY);
+
+        String key = S3Repository.getValue(repositorySettings,
+            S3Repository.Repository.KEY_SETTING, S3Repository.Repositories.KEY_SETTING);
+        String secret = S3Repository.getValue(repositorySettings,
+            S3Repository.Repository.SECRET_SETTING, S3Repository.Repositories.SECRET_SETTING);
+
+        AWSCredentialsProvider credentialsProvider = InternalAwsS3Service.buildCredentials(logger, key, secret);
+
+        try {
+            AWSCredentials credentials = credentialsProvider.getCredentials();
+            assertThat("We should have raised an exception as no credentials are available", expectedCredentials, notNullValue());
+
+            assertThat(credentials.getAWSAccessKeyId(), is(expectedCredentials.getAWSAccessKeyId()));
+            assertThat(credentials.getAWSSecretKey(), is(expectedCredentials.getAWSSecretKey()));
+        } catch (AmazonClientException e) {
+            assertThat("We should not have raised an exception as we have credentials", expectedCredentials, nullValue());
```

Comment:
@rjernst I added a new commit which I expect make the test more readable/useful.